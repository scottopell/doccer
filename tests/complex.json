{
  "root": 1,
  "index": {
    "1": {
      "id": 1,
      "crate_id": 0,
      "name": "complex",
      "visibility": "public",
      "docs": "Complex fixture for testing doccer\n\nThis crate contains more complex Rust constructs to validate\nadvanced parsing and rendering functionality.",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"module": {"items": [2, 3, 4, 5]}}
    },
    "2": {
      "id": 2,
      "crate_id": 0,
      "name": "HttpClient",
      "visibility": "public",
      "docs": "A client for making HTTP requests",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"struct": {"fields": [], "impls": [6, 7, 8]}}
    },
    "3": {
      "id": 3,
      "crate_id": 0,
      "name": "Method",
      "visibility": "public",
      "docs": "HTTP request methods",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"enum": {"variants": [9, 10, 11, 12, 13]}}
    },
    "4": {
      "id": 4,
      "crate_id": 0,
      "name": "Protocol",
      "visibility": "public",
      "docs": "Protocol for network communications",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"trait": {"items": [14, 15, 16]}}
    },
    "5": {
      "id": 5,
      "crate_id": 0,
      "name": "sub_module",
      "visibility": "public",
      "docs": "A submodule with additional types",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"module": {"items": [17, 18]}}
    },
    "6": {
      "id": 6,
      "crate_id": 0,
      "name": null,
      "visibility": "public",
      "docs": null,
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"impl": {"for": {"resolved_path": {"path": "HttpClient"}}, "items": [19, 20, 21]}}
    },
    "7": {
      "id": 7,
      "crate_id": 0,
      "name": null,
      "visibility": "public",
      "docs": null,
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"impl": {"trait": {"path": "std::fmt::Debug"}, "for": {"resolved_path": {"path": "HttpClient"}}, "is_synthetic": true}}
    },
    "8": {
      "id": 8,
      "crate_id": 0,
      "name": null,
      "visibility": "public",
      "docs": null,
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"impl": {"trait": {"path": "Protocol"}, "for": {"resolved_path": {"path": "HttpClient"}}, "items": [22, 23, 24]}}
    },
    "9": {
      "id": 9,
      "crate_id": 0,
      "name": "Get",
      "visibility": "public",
      "docs": "HTTP GET method",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"variant": {"kind": "plain"}}
    },
    "10": {
      "id": 10,
      "crate_id": 0,
      "name": "Post",
      "visibility": "public",
      "docs": "HTTP POST method",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"variant": {"kind": "plain"}}
    },
    "11": {
      "id": 11,
      "crate_id": 0,
      "name": "Put",
      "visibility": "public",
      "docs": "HTTP PUT method",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"variant": {"kind": "plain"}}
    },
    "12": {
      "id": 12,
      "crate_id": 0,
      "name": "Delete",
      "visibility": "public",
      "docs": "HTTP DELETE method",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"variant": {"kind": "plain"}}
    },
    "13": {
      "id": 13,
      "crate_id": 0,
      "name": "Custom",
      "visibility": "public",
      "docs": "Custom HTTP method",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"variant": {"kind": {"tuple": [25]}}}
    },
    "14": {
      "id": 14,
      "crate_id": 0,
      "name": "Error",
      "visibility": "public",
      "docs": "Associated error type",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"assoc_type": {"bounds": [{"trait_bound": {"trait": {"path": "std::error::Error"}}}]}}
    },
    "15": {
      "id": 15,
      "crate_id": 0,
      "name": "connect",
      "visibility": "public",
      "docs": "Connect to the endpoint",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"function": {"sig": {"inputs": [["self", {"borrowed_ref": {"is_mutable": true, "type": {"generic": "Self"}}}], ["url", {"resolved_path": {"path": "String"}}]], "output": {"resolved_path": {"path": "Result", "args": {"angle_bracketed": {"args": [{"type": {"tuple": []}}, {"type": {"qualified_path": {"name": "Error"}}}]}}}}}}}
    },
    "16": {
      "id": 16,
      "crate_id": 0,
      "name": "disconnect",
      "visibility": "public",
      "docs": "Disconnect from the endpoint",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"function": {"sig": {"inputs": [["self", {"borrowed_ref": {"is_mutable": true, "type": {"generic": "Self"}}}]], "output": {"tuple": []}}}}
    },
    "17": {
      "id": 17,
      "crate_id": 0,
      "name": "Response",
      "visibility": "public",
      "docs": "HTTP response",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"struct": {"fields": [26, 27, 28], "impls": [29]}}
    },
    "18": {
      "id": 18,
      "crate_id": 0,
      "name": "Headers",
      "visibility": "public",
      "docs": "HTTP headers collection",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"struct": {"fields": [], "impls": []}}
    },
    "19": {
      "id": 19,
      "crate_id": 0,
      "name": "new",
      "visibility": "public",
      "docs": "Creates a new HTTP client",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"function": {"sig": {"inputs": [], "output": {"generic": "Self"}}}}
    },
    "20": {
      "id": 20,
      "crate_id": 0,
      "name": "request",
      "visibility": "public",
      "docs": "Makes an HTTP request",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"function": {"sig": {"inputs": [["self", {"borrowed_ref": {"type": {"generic": "Self"}}}], ["method", {"resolved_path": {"path": "Method"}}], ["url", {"resolved_path": {"path": "String"}}]], "output": {"resolved_path": {"path": "Result", "args": {"angle_bracketed": {"args": [{"type": {"resolved_path": {"path": "sub_module::Response"}}}, {"type": {"resolved_path": {"path": "String"}}}]}}}}}}}
    },
    "21": {
      "id": 21,
      "crate_id": 0,
      "name": "with_timeout",
      "visibility": "public",
      "docs": "Sets a timeout for requests",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"function": {"sig": {"inputs": [["self", {"generic": "Self"}], ["seconds", {"primitive": "u64"}]], "output": {"generic": "Self"}}}}
    },
    "22": {
      "id": 22,
      "crate_id": 0,
      "name": "Error",
      "visibility": "public",
      "docs": null,
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"assoc_type": {"type": {"resolved_path": {"path": "HttpError"}}}}
    },
    "23": {
      "id": 23,
      "crate_id": 0,
      "name": "connect",
      "visibility": "public",
      "docs": "Implement connect for HTTP",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"function": {"sig": {"inputs": [["self", {"borrowed_ref": {"is_mutable": true, "type": {"generic": "Self"}}}], ["url", {"resolved_path": {"path": "String"}}]], "output": {"resolved_path": {"path": "Result", "args": {"angle_bracketed": {"args": [{"type": {"tuple": []}}, {"type": {"resolved_path": {"path": "HttpError"}}}]}}}}}}}
    },
    "24": {
      "id": 24,
      "crate_id": 0,
      "name": "disconnect",
      "visibility": "public",
      "docs": "Implement disconnect for HTTP",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"function": {"sig": {"inputs": [["self", {"borrowed_ref": {"is_mutable": true, "type": {"generic": "Self"}}}]], "output": {"tuple": []}}}}
    },
    "25": {
      "id": 25,
      "crate_id": 0,
      "name": null,
      "visibility": "public",
      "docs": null,
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"struct_field": {"resolved_path": {"path": "String"}}}
    },
    "26": {
      "id": 26,
      "crate_id": 0,
      "name": "status",
      "visibility": "public",
      "docs": "HTTP status code",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"struct_field": {"primitive": "u16"}}
    },
    "27": {
      "id": 27,
      "crate_id": 0,
      "name": "headers",
      "visibility": "public",
      "docs": "Response headers",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"struct_field": {"resolved_path": {"path": "Headers"}}}
    },
    "28": {
      "id": 28,
      "crate_id": 0,
      "name": "body",
      "visibility": "public",
      "docs": "Response body",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"struct_field": {"resolved_path": {"path": "String"}}}
    },
    "29": {
      "id": 29,
      "crate_id": 0,
      "name": null,
      "visibility": "public",
      "docs": null,
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"impl": {"for": {"resolved_path": {"path": "Response"}}, "items": [30]}}
    },
    "30": {
      "id": 30,
      "crate_id": 0,
      "name": "is_success",
      "visibility": "public",
      "docs": "Checks if the response status is a success code",
      "links": {},
      "attrs": [],
      "deprecation": null,
      "inner": {"function": {"sig": {"inputs": [["self", {"borrowed_ref": {"type": {"generic": "Self"}}}]], "output": {"primitive": "bool"}}}}
    }
  }
}