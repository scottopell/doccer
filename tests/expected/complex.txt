# Crate: complex

Version: 0.1.0

Complex fixture for testing doccer

This crate combines all Rust features to provide a comprehensive
test of doccer's parsing and rendering capabilities.

# Examples

```rust
use complex::storage::*;
let mut storage = Storage::new();
storage.insert("key", "value");
```

  /// A macro for creating formatted messages
  ///
  /// # Examples
  ///
  /// ```
  /// let msg = format_message!("Hello", "World");
  /// assert_eq!(msg, "Hello: World");
  /// ```
  macro_rules! format_message($prefix:expr, $content:expr)

  /// Storage and data management
  pub mod storage

    /// Specialized cache for strings
    pub mod string_cache

      /// A string cache with complex operations
      pub struct StringCache {

        /// Creates a new string cache
        pub fn new() -> Self

        /// Adds a string to the cache
        pub fn add(&mut self, key: String, content: String)
      }

      /// A cached string with metadata
      pub struct CachedString {
      }

      /// Implementation of Clone trait for CachedString
      impl Clone for CachedString {

        fn clone(&self) -> CachedString
      }

      /// Implementation of Debug trait for CachedString
      impl Debug for CachedString {

        fn fmt(&self, f: &mut $crate::fmt::Formatter<'_>) -> $crate::fmt::Result
      }

    /// A generic storage container with complex constraints
    pub struct Storage<K, V> where K: Clone + Debug + PartialEq + std::hash::Hash, V: Clone + Debug {

      /// Creates a new storage with default capacity
      ///
      /// # Examples
      ///
      /// ```
      /// let storage: Storage<String, i32> = Storage::new();
      /// ```
      pub fn new() -> Self

      /// Creates a new storage with specified capacity
      pub fn with_capacity(capacity: usize) -> Self

      /// Inserts a key-value pair
      pub fn insert(&mut self, key: K, value: V) -> Option<V>

      /// Gets a value by key
      pub fn get(&self, key: &K) -> Option<&V>
    }

    /// A trait for cacheable items
    pub trait Cacheable<K> where K: Clone {

      /// The cache key type
      type Key: Clone + Debug

      /// Gets the cache key for this item
    fn cache_key(&self) -> Self::Key

      /// Validates if the item can be cached
    fn can_cache(&self) -> bool
    }

    /// Cache implementation with lifetime parameters
    pub struct Cache<'a, T: Cacheable> {

      /// Creates a new cache
      pub fn new(data: &'a [T]) -> Self
    }

  /// Network operations and protocols
  pub mod network

    /// Protocol definitions
    pub mod protocol

      /// A generic protocol handler
      pub trait Protocol<Req, Resp> {

        /// The error type for this protocol
        type Error: std::error::Error

        /// Processes a request
        fn handle(&mut self, request: Req) -> Result<Resp, Self::Error>
      }

      /// HTTP protocol implementation
      pub struct Http {
      }

      /// Implementation of Protocol trait for Http
      impl Protocol<HttpRequest, HttpResponse> for Http {

        type Error = HttpError

        fn handle(&mut self, request: HttpRequest) -> Result<HttpResponse, Self::Error>
      }

      /// HTTP request
      pub struct HttpRequest {
      }

      /// Implementation of Protocol trait for Http
      impl Protocol<HttpRequest, HttpResponse> for Http {

        type Error = HttpError

        fn handle(&mut self, request: HttpRequest) -> Result<HttpResponse, Self::Error>
      }

      /// HTTP response
      pub struct HttpResponse {
      }

      /// Implementation of Protocol trait for Http
      impl Protocol<HttpRequest, HttpResponse> for Http {

        type Error = HttpError

        fn handle(&mut self, request: HttpRequest) -> Result<HttpResponse, Self::Error>
      }

      /// HTTP error
      pub struct HttpError {
      }

      /// Implementation of Debug trait for HttpError
      impl Debug for HttpError {

        fn fmt(&self, f: &mut $crate::fmt::Formatter<'_>) -> $crate::fmt::Result
      }

      /// Implementation of Display trait for HttpError
      impl Display for HttpError {

        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result
      }

      /// Implementation of Error trait for HttpError
      impl Error for HttpError

  /// Mathematical operations and utilities
  pub mod math

    /// Constants for mathematical operations
    pub mod constants

      /// Mathematical constant Ï€
      pub const PI: f64

      /// Mathematical constant e
      pub const E: f64

      /// Golden ratio
      pub const PHI: f64

    /// A point in 2D space
    pub struct Point<T: Copy> {

      /// Creates a new point
      pub fn new(x: T, y: T) -> Self

      /// Adds two points together
      pub fn add(self, other: Self) -> Self
    }

    /// Implementation of Debug trait for Point
    impl Debug for Point<T> {

    fn fmt(&self, f: &mut $crate::fmt::Formatter<'_>) -> $crate::fmt::Result
    }

    /// Implementation of Clone trait for Point
    impl Clone for Point<T> {

    fn clone(&self) -> Point<T>
    }

    /// Implementation of Copy trait for Point
    impl Copy for Point<T>

    /// Implementation of PartialEq trait for Point
    impl PartialEq for Point<T> {

    fn eq(&self, other: &Point<T>) -> bool
    }

# Re-exports

  /// Re-exports for convenience
  pub use math::Point
  pub use network::protocol::Http
  pub use network::protocol::Protocol
  pub use storage::Storage